### PR TITLE
[master] Cleanup of fixes for chainlink

### PR DESCRIPTION
**Pull request by n-hutton**
Fixes for chainlink, comments inline:
- Subscription bug
- Prev hash not set
- Block JSON fields not parseable

There are not CI tests for these, it would be good if there were the following tests:
- Blocks have prevHash set correctly
- Subscription to newHeads works, and it works if you subscribe multiple times (all subscribers recieve the block)

### CODE REVIEW

File: src/libServer/IsolatedServer.cpp [hunk 1]
```diff
@@ -1181,9 +1181,13 @@ TxBlock IsolatedServer::GenerateTxBlock() {
...
   }
 
+  auto const prevTxBlock = m_sharedMediator.m_txBlockChain.GetLastBlock();
+
+  auto const prevHash = m_blocknum == 0 ? BlockHash(): prevTxBlock.GetBlockHash();
+
   TxBlockHeader txblockheader(0, m_currEpochGas, 0, m_blocknum,
                               TxBlockHashSet(), numtxns, m_key.first,
-                              TXBLOCK_VERSION);
+                              1, TXBLOCK_VERSION, CommitteeHash(), prevHash);
```
Comment (n-hutton) [inline src/libServer/IsolatedServer.cpp:1190]: this was actually incedentally incorrectly setting the version as the ds block number here

File: src/libServer/IsolatedServer.cpp [hunk 2]
```diff
@@ -1181,9 +1181,13 @@ TxBlock IsolatedServer::GenerateTxBlock() {
     m_txnBlockNumMap[m_blocknum].clear();
   }
 
+  auto const prevTxBlock = m_sharedMediator.m_txBlockChain.GetLastBlock();
+
+  auto const prevHash = m_blocknum == 0 ? BlockHash(): prevTxBlock.GetBlockHash();
```
Comment (n-hutton) [inline src/libServer/IsolatedServer.cpp:1186]: Set the prev hash in blocks (iso server only)

File: src/libServer/JSONConversion.cpp [hunk 3]
```diff
@@ -160,8 +160,8 @@ const Json::Value JSONConversion::convertTxBlocktoEthJson(
   // Required by ethers
   retJson["extraData"] = "0x";
   retJson["nonce"] = "0x0";
-  retJson["receiptsRoot"] = "0x";
-  retJson["transactionsRoot"] = "0x0";
+  retJson["receiptsRoot"] = "0x0000000000000000000000000000000000000000000000000000000000000000";
```
Comment (n-hutton) [inline src/libServer/JSONConversion.cpp:163]: chainlink unable to parse non hash-length strings, so make them full length. I think this should be fine, no client actually uses these fields I believe

File: tests/EvmAcceptanceTests/hardhat.config.ts [hunk 4]
```diff
@@ -113,13 +113,15 @@ const config: HardhatUserConfig = {
       miningState: false
     },
     testnet: {
-      url: "https://devnetnh-l2api.dev.z7a.xyz",
-      websocketUrl: "wss://devnetnh-l2api.dev.z7a.xyz",
+      url: "https://testnet-receipts-l2api.dev.z7a.xyz",
```
Comment (n-hutton) [inline tests/EvmAcceptanceTests/hardhat.config.ts:116]: convenience

File: src/libEth/filters/SubscriptionsImpl.h [hunk 5]
```diff
@@ -70,14 +71,32 @@ class SubscriptionsImpl {
...
     Id id = 0;
 
-    /// True if this conn subscribed to pending txns
-    bool subscribedToPendingTxns = false;
+    // populated if this conn subscribed to pending txns.
+    // since a single connection could have multiple subscriptions, we need to
+    // keep a set of subscription IDs here. They are non-overlapping
+    // with subscribedToNewHeads
+    //bool subscribedToPendingTxns = false;
+    std::unordered_set<uint64_t> subscribedToPendingTxn;
```
Comment (n-hutton) [inline src/libEth/filters/SubscriptionsImpl.h:79]: So, the same subscription ID was being used for every connection to newHeads. This is completely fine until there are multiple newHeads subscriptions in the same single connection. This is a bit of a corner case.

Anyway, to solve this, connectinos now keep a set of subscription IDs which are unique and not re-used for subscription types (hewHeads, pendingTxn). This has been confirmed to work with the chainlink software.

**Review by Steve-White-UK — APPROVED** [review]
OK

**Pull request merged**